### PR TITLE
Check pipeline layout limits

### LINKED ISSUE #585: Check limits of shaders
When a shader module is loaded, its limits should be tested against the physical device's limits.

For example: the compute shader local size.

### PR DESCRIPTION
**Pull request by tomaka**
Fix #585

### CODE REVIEW

File: vulkano/src/descriptor/pipeline_layout/limits_check.rs [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+    let mut num_storage_images = Counter::default();
+    let mut num_input_attachments = Counter::default();
+
+    for set in 0 .. desc.num_sets() {
+        for binding in 0 .. desc.num_bindings_in_set(set).unwrap() {
+            let descriptor = desc.descriptor(set, binding).unwrap();
+
+            match descriptor.ty.ty().expect("Not implemented yet") {        // TODO:
+                DescriptorType::Sampler => {
+                    num_resources.increment(descriptor.array_count, &descriptor.stages);
```
Comment (nicokoch) [inline vulkano/src/descriptor/pipeline_layout/limits_check.rs]: This line in each match arm can be removed and added once before the match

**Comment by tomaka** [comment]
I wonder if `maxUniformBufferRange` and `maxStorageBufferRange` should also be checked.
The specs say that these limits must be respected when binding a buffer. In theory you could bind only a certain range of a uniform buffer if you know that only this range will be accessed by the shader, in which case what you're doing is valid.

**Comment by tomaka** [comment]
Note that vulkano already checks maxUniformBufferRange and maxStorageBufferRange when binding the buffer.

**Comment by nicokoch** [comment]
IMO they should be checked when binding the buffer, not when creating the layout, as the specs say and as vulkano already does. So everything is fine like it is

**Pull request merged**